scipy/stats/_morestats.py

```diff
@@ -1179,6 +1186,14 @@ def boxcox_normmax(x, brack=None, method='pearsonr', optimizer=None):
 
         See the example below or the documentation of
         `scipy.optimize.minimize_scalar` for more information.
+    ymax : float, optional
+        The unconstrained optimal transform parameter may cause Box-Cox
+        transformed data to have extreme magnitude or even overflow.
+        This parameter constrains MLE optimization such that the magnitude
+        of the transformed `x` does not exceed `ymax`. The default is
+        the maximum value of the input dtype. If set to infinity,
+        `boxcox_normmax` returns the unconstrained optimal lambda.
+        Ignored when ``method='pearsonr'``.
```

Comment:
I think user can set `ymax` to 1e100 for example or a smaller value to avoid overflow on new data.

Reply:
That's possible yes, but I also believe that users should be able to count on the transform not overflowing for most applications. And the current default for `ymax` would not prevent overflow for most applications. I'm not sure there even is a good default value for `ymax`, as it depends on the largest value you expect to transform – hence the suggestion of taking `xmax` into account (explicitly or implicitly).

Reply:
`xmax` is a reasonable idea.  It can be added later if desired, but if we think we're going to add it, we should discuss the idea of specifying both `ymax` and `xmax` in the same argument. That might change the name of the argument we add.

> And the current default

IIRC, the safety factor in YJ takes into account the dispersion of the current `x`, but it is not guaranteed to prevent overflow either.

We can do something similar here, but I don't think what is here is wrong.

Reply:
If this PR doesn't add an `xmax` or similar, then I'm in favour of applying a safety factor on the extremal values of `x` of say 20 to help avoid unexpected overflow. This would not invalidate the promise that `y = boxcox(x, lmbda)` would be no larger than `ymax` in absolute value.

On the topic of the default value for `ymax`: if we don't reserve half of the exponent space, users may not be able to compute the variance of the result by default. That's a scenario that's not unimaginable if for instance the user would want to z-score the result.

Reply:
I'll let you two discuss the default. Are people really interested in transforming moderate data and getting transformed data with absurdly large magnitudes? Should the default depend on the magnitude of the input?

I don't have experience *using* this function, so I defer to others here.

Reply:
There is often a difference between the data to learn/fit and the data to apply it. The 2000 vs 3000 does not seem unrealistic to me. Therefore, I‘d favor an explicit `xmax`. I would advice against some random choice of a safety margin.

Reply:
> The user sets an xmax in addition to ymax with which they can indicate the maximum value they expect to observe for future transformation. This is explicit, but it further complicates the API.

If we add `xmax`, what is the default value for it or leave if as `None`? If user set `xmax` to be 20 times the maximum of current data,  overflow might still occur on future data.

> That's a scenario that's not unimaginable if for instance the user would want to z-score the result.

User can still z-score the result. 
```python
import numpy as np
from scipy.special import boxcox


x = [2003.0, 1950.0, 1997.0, 2000.0, 2009.0]
y = boxcox(x, 90)

ymean = np.mean(y)
yvar = np.var(y / ymean)
y_zscore = (y - ymean) / ymean / np.sqrt(yvar)

print(y_zscore) # [ 0.47912058 -1.78257126 -0.10864857  0.16564861  1.24645064]
print(np.mean(y_zscore)) # 0.0
print(np.var(y_zscore)) # 0.9999999999999997
```

Reply:
If we add a safety factor to x_treme, I guess we need to do the following (since Box-Cox support strictly positive data), right?
- `x_treme` *= 20  if `x_treme` == `xmax` (positive overflow)
- `x_treme` /= 20  if `x_treme` == `xmin` (negative overflow)

Then, back to my previous question, adding `xmax` seems not help with negative overflow here.

Reply:
I suppose it would be a tuple `xbounds` then.